src/main/java/io/cdap/plugin/marketo/source/batch/MarketoRecordReader.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * RecordReader implementation, which reads events from Marketo api.
+ */
+public class MarketoRecordReader extends RecordReader<NullWritable, Map<String, String>> {
+  private static final Logger LOG = LoggerFactory.getLogger(MarketoRecordReader.class);
+  private static final Gson GSON = new GsonBuilder().create();
+  private Map<String, String> current = null;
+  private Iterator<CSVRecord> iterator = null;
+  private String beginDate;
+  private String endDate;
+
+  public MarketoRecordReader(String beginDate, String endDate) {
+    this.beginDate = beginDate;
+    this.endDate = endDate;
+  }
+
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) throws IOException {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(MarketoInputFormatProvider.PROPERTY_CONFIG_JSON);
+    MarketoReportingSourceConfig config = GSON.fromJson(configJson, MarketoReportingSourceConfig.class);
+
+    Marketo marketo = config.getMarketo();
+
+    DateRange dateRange = new DateRange(beginDate, endDate);
+    AbstractBulkExportJob job;
+    switch (config.getReportType()) {
+      case LEADS:
+        List<String> leadsFields = config.getSchema().getFields().stream()
+          .map(Schema.Field::getName).collect(Collectors.toList());
+        LeadsExportRequest.ExportLeadFilter leadsFilter = LeadsExportRequest.ExportLeadFilter.builder()
+          .createdAt(dateRange).build();
+        job = marketo.exportLeads(new LeadsExportRequest(leadsFields, leadsFilter));
+        break;
+      case ACTIVITIES:
+        ExportActivityFilter activitiesFilter = ExportActivityFilter.builder()
+          .createdAt(dateRange).build();
+        job = marketo.exportActivities(new ActivitiesExportRequest(Marketo.ACTIVITY_FIELDS, activitiesFilter));
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid report type " + config.getReportType());
+    }
+
+    LOG.info("BULK EXPORT JOB - job '{}' has date range '{}'", job.getJobId(), dateRange);
+
+    // TODO handle possible concurrent issues here, another mapper can take our slot
+    // wait for 10 minutes for available slot and enqueue job
+    marketo.onBulkExtractQueueAvailable(job::enqueue, 60 * 10);
+
+    job.waitCompletion();
+
+    String data = job.getFile();
+    // TODO stream here
+    CSVParser parser = CSVFormat.DEFAULT.withHeader().parse(new StringReader(data));
+    iterator = parser.iterator();
+  }
+
+  @Override
+  public boolean nextKeyValue() {
+    if (iterator.hasNext()) {
+      current = iterator.next().toMap();
+      LOG.debug("Got record '{}'", current.toString());
```

Comment:
I remember when Google guys reviewed my PRs they always told me to remove debug logs, since it floods the full log too much with millions of records. So maybe don't add them in places which are called too often.

Reply:
Leftovers from debugging, removed.